### PR TITLE
How to integrate a new minigame

### PR DESCRIPTION
**Pull request by maexled**
Closes Gamify-IT/issues#312

### CODE REVIEW

**Comment by Aaron2000119** [comment]
you probably have to change it once route validation is added in the overworld-backend. I don't know if you want to wait till then or if you want to merge it now and change it later.

File: dev-manuals/services/integrate-new-minigame.md [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+export async function get<YOUR_MINIGAME_NAME>Config(
+  id: string
+): Promise<AxiosResponse> {
+  return axios.get(`${config.<YOUR_MINIGAME_NAME>ApiUrl}/configurations/${id}`);
+}
+```
+
+### Create a configuration modal
+
+Create a new configuration modal under the paclage `components.EditMinigameModals`.
```
Comment (delvh) [inline dev-manuals/services/integrate-new-minigame.md:169]:
```suggestion
Create a new configuration modal under the package `components.EditMinigameModals`.  
```

File: dev-manuals/services/integrate-new-minigame.md [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+We can't document exactly what the code should look like, as it varies a lot, but you can orient yourself on the existing edit components. It is important to say that the minigame configurations should be loaded and updated in the component. When a modal gets closed call an `closedModal emit` and when a configuration gets updated call a `updateMinigameConfiguration emit`.
```
Comment (delvh) [inline dev-manuals/services/integrate-new-minigame.md:170]:
```suggestion
We can't document exactly what the code should look like, as it varies a lot, but you can orient yourself on the existing edit components.  
It is important that the minigame configurations are loaded and updated in the component.  
When a modal gets closed, emit `closedModal` and when a configuration gets updated emit `updateMinigameConfiguration`.
```

File: dev-manuals/services/integrate-new-minigame.md [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+
+The minigames edit view is in the class `MinigameTasksView.vue` under the package `views`.
+
+Create a new variable:
+```typescript
+const show<YOUR_MINIGAME_NAME>Modal = ref(false);
+```
+
+Add in the editMinigameConfiguration method your minigame:
```
Comment (delvh) [inline dev-manuals/services/integrate-new-minigame.md:179]:
```suggestion
Add your minigame in the `editMinigameConfiguration` method:
```

File: dev-manuals/services/integrate-new-minigame.md [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+```typescript
+case Minigame.<YOUR_MINIGAME_NAME>:
+    show<YOUR_MINIGAME_NAME>Modal.value = true;
+    break;
+```
+
+Add in the closedEditModal method following at the bottom:
```
Comment (delvh) [inline dev-manuals/services/integrate-new-minigame.md:186]:
```suggestion
Append the following at the bottom of the `closedEditModal` method:
```

File: dev-manuals/services/integrate-new-minigame.md [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+import { saveAs } from "file-saver";
+import { arrayOf, defaultValue, object, string, int } from "checkeasy"; // depends on your configuration attributes
+import { importConfiguration } from "@/ts/import-configuration";
+import ImportExportConfiguration from "@/components/ImportExportConfiguration.vue";
+```
+
+To make a configuration downloadable, create the method downloadConfiguration:
+```typescript
+function downloadConfiguration() {
+  const { ["id"]: unused, ...clonedConfiguration } = configuration.value;
```
Comment (delvh) [inline dev-manuals/services/integrate-new-minigame.md:215]: Isn't there the backend call missing?

File: dev-manuals/services/integrate-new-minigame.md [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+```typescript
+<Edit<YOUR_MINIGAME_NAME>Modal
+    :showModal="show<YOUR_MINIGAME_NAME>Modal"
+    :minigame="editedMinigame"
+    @updateMinigameConfiguration="updateMinigameConfiguration"
+    @closedModal="closedEditModal"
+  />
+```
+
+#### Integrate import/export in the edit modal
```
Comment (delvh) [inline dev-manuals/services/integrate-new-minigame.md:201]:
```suggestion
#### Add configuration import/ export
```

File: dev-manuals/services/integrate-new-minigame.md [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+
+Finally the component to import and export a file needs to be added, which displays a import and an export button. Simply add the following code in your modal:
+```typescript
+<ImportExportConfiguration
+    @export="downloadConfiguration"
+    @importFile="importFile"
+/>
+```
+
+## Send game results from the minigame backend to the overworld backend
```
Comment (delvh) [inline dev-manuals/services/integrate-new-minigame.md:266]:
```suggestion
## Backend-to-backend communication
```

File: dev-manuals/services/integrate-new-minigame.md [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+
+The minigame backend must send a score after every player's game run to the overworld backend after a player's playthrough. The score must be between 0 and 100, where 100 is a completely correct run and 0 is a completely wrong run. The score must be calculated by the minigame backend itself.
```
Comment (delvh) [inline dev-manuals/services/integrate-new-minigame.md:268]:
```suggestion
The minigame backend must send a score after every player's game run to the overworld backend after a player's playthrough. The score must be a natural number between `0` and `100`, where `100` means "no further improvements possible"  and `0` means "everything is wrong".  
The minigame backend is responsible for calculating this score, the overworld backend simply stores it. 
```

File: dev-manuals/services/integrate-new-minigame.md [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+    @export="downloadConfiguration"
+    @importFile="importFile"
+/>
+```
+
+## Send game results from the minigame backend to the overworld backend
+
+The minigame backend must send a score after every player's game run to the overworld backend after a player's playthrough. The score must be between 0 and 100, where 100 is a completely correct run and 0 is a completely wrong run. The score must be calculated by the minigame backend itself. 
+
+Create the OverworldResultDTO:
```
Comment (delvh) [inline dev-manuals/services/integrate-new-minigame.md:270]:
```suggestion
The following shows how to do it in Java with `Spring`, `Lombok`, and `FeignClient`.

First, create the `OverworldResultDTO`:
```

**Review by Gr33ndev — APPROVED** [review]
see @delvh & @Aaron2000119 comments, everything else looks good to me

**Review by robojones — APPROVED** [review]
lgtm. You should probably apply the suggestions from @delvh

**Comment by maexled** [comment]
> you probably have to change it once route validation is added in the overworld-backend. I don't know if you want to wait till then or if you want to merge it now and change it later.

What has to be changed?

File: dev-manuals/services/integrate-new-minigame.md [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+```typescript
+import { saveAs } from "file-saver";
+import { arrayOf, defaultValue, object, string, int } from "checkeasy"; // depends on your configuration attributes
+import { importConfiguration } from "@/ts/import-configuration";
+import ImportExportConfiguration from "@/components/ImportExportConfiguration.vue";
+```
+
+To make a configuration downloadable, create the method downloadConfiguration:
+```typescript
+function downloadConfiguration() {
```
Comment (delvh) [inline dev-manuals/services/integrate-new-minigame.md:223]:
```suggestion
To make a configuration downloadable, create the method `downloadConfiguration`:
```typescript
// configuration is expected to be a ref to the current configuration
function downloadConfiguration() {
```

**Review by delvh — APPROVED** [review]

File: dev-manuals/services/integrate-new-minigame.md (hunk 10 above)
Comment (delvh) [reply dev-manuals/services/integrate-new-minigame.md:223]: That was a bit too fast. There is still an open suggestion.

**Pull request merged**